packages/lesswrong/lib/recommendations.js

```diff
@@ -0,0 +1,76 @@
+import { addGraphQLResolvers, addGraphQLQuery } from 'meteor/vulcan:core';
+import { Posts } from './collections/posts';
+
+const getLoggedOutRecommendations = async ({count}) => {
+  let topPosts = await Posts.find(
+    {},
+    {
+      limit:count,
+      sort: {baseScore: -1}
+    }
+  ).fetch();
+  return topPosts;
+}
+
+const getRecommendations = async ({count, currentUser}) => {
+  if (!currentUser) {
+    return await getLoggedOutRecommendations({count});
+  }
+  
+  const unreadTopPostsMeta = await Posts.aggregate([
+    { $match: {
+      status: Posts.config.STATUS_APPROVED,
+      isFuture: false,
+      baseScore: {$gt: 30},
+      draft: false,
+      unlisted: false,
+      authorIsUnreviewed: false,
+      groupId: {$exists: false},
+    } },
+    { $project: {_id:1, baseScore:1} },
+    { $sort: {baseScore:-1} },
+    
+    { $lookup: {
+      from: "lwevents",
+      let: { documentId: "$_id", },
+      pipeline: [
+        { $match: {
+          name: "post-view",
+          userId: currentUser._id,
+        } },
+        { $match: { $expr: {
+          $and: [
+            {$eq: ["$documentId", "$$documentId"]},
+          ]
+        } } },
+        { $limit: 1},
+      ],
+      as: "views",
+    } },
+    
+    { $match: {
+      "views": {$size:0}
```

Comment:
I would have compared `views` to the empty array, instead of checking it's length. Though it shouldn't make much of a difference either way.

Reply:
MongoDB has a lot of weird special cases with arrays, which I don't really trust my understanding of.